lib/python/qmk/cli/compile.py

```diff
@@ -28,6 +29,48 @@ def compile(cli):
     If --keyboard and --keymap are provided this command will build a firmware based on that.
 
     """
+    # Set CWD as directory command was issued from
+    cwd = os.environ['ORIG_CWD']
+    qmk_path = os.getcwd()
+    current_folder = os.path.basename(cwd)
+    # Initialize boolean to check for being in a keyboard directory and initialize keyboard string
+    in_keyboard = False
+    in_layout = False
+    keyboard = ""
+    keymap = ""
+    user_keymap = ""
+    user_keyboard = ""
+
+    # Set path for '/keyboards/' directory
+    keyboards_path = os.path.join(qmk_path , "keyboards")
+    layouts_path = os.path.join(qmk_path, "layouts")
+
+    # If below 'keyboards' and not in 'keyboards' or 'keymaps', get current keyboard name
+    if cwd.startswith(keyboards_path):
+        if current_folder != "keyboards" and current_folder != "keymaps":
+            if os.path.basename(os.path.abspath(os.path.join(cwd, ".."))) == "keymaps":
+                # If in a keymap folder, set relative path, get everything before /keymaps, and the keymap name
+                relative_path = cwd[len(keyboards_path):][1:]
+                keyboard = str(relative_path).split("/keymaps", 1)[0]
+                keymap = str(relative_path.rsplit("/", 1)[-1])
+            else:
+                keyboard = str(cwd[len(keyboards_path):])[1:]
+
+            in_keyboard= True
+
+    # If in layouts dir
+    if cwd.startswith(layouts_path):
+        if current_folder != "layouts":
+            layout = str(cwd[len(layouts_path):])[1:]
+            in_layout=True
+
+    # If user keyboard/keymap or compile keyboard/keymap are supplied, assign those
+    if cli.config.compile.keyboard or cli.config.user.keyboard:
+        user_keyboard = cli.config.user.keyboard if cli.config.user.keyboard else cli.config.compile.keyboard
+    if cli.config.compile.keymap or cli.config.user.keymap:
+        if not in_layout:
```

Comment:
And the same goes for the keymap:

```suggestion

```

Reply:
This would not solve the comment from above here: https://github.com/qmk/qmk_firmware/pull/6884#pullrequestreview-306268697. 

Removing this check will set `user_keyboard` and `user_keymap` based on the config, which would remove the layout directory context

Reply:
Hmm, you are right.
Imho, we should work out the following precedence order: command line > directory context > config

Reply:
> Imho, we should work out the following precedence order: command line > directory context > config

Yes, this is the correct precedence.